src/quicksort.rs

```diff
@@ -17,41 +17,32 @@ fn quicksort<T: Ord>(slice: &mut [T]) {
     let (pivot, rest) = slice.split_first_mut().expect("slice is non-empty");
     let mut left = 0;
     let mut right = rest.len() - 1;
-    while left <= right {
+    // if right <= 0, it's done
+    while left <= right && right > 0 {
         if &rest[left] <= pivot {
             // already on the correct side
             left += 1;
-        } else if &rest[right] > pivot {
+        } else if &rest[right] > pivot && right > 0 {
```

Comment:
I don't think this conditional change is necessary? After all, we won't _get_ here with the new `right > 0` in the `while` conditional.